### ISSUE TITLE
force rstan to use makeForkCluster on Linux

### ISSUE THREAD

**Issue by wookietreiber** [stan-dev/rstan#354]
**rstan on Linux** defaults to `makePSOCKcluster` (possibly because rstan is using `makeCluster` which in turn defaults to `makePSOCKcluster` on Linux). The PSOCK cluster does not work that well on Linux, e.g. because resources are not cleaned up very well. In my experience, `makeForkCluster` works much better on Linux.

Is there a way to enforce the usage of `makeForkCluster`?

---

Besides the cleanup problems, we are now also getting sporadic errors, so it would be really nice if we could enforce the usage of `makeForkCluster`:

```
Error in socketConnection("localhost", port = port, server = TRUE, blocking = TRUE,  :

  cannot open the connection

Calls: stan ... makePSOCKcluster -> newPSOCKnode -> socketConnection

In addition: Warning messages:

1: In FUN(X[[i]], ...) : data with name spname is not numeric and not used

2: In socketConnection("localhost", port = port, server = TRUE, blocking = TRUE,  :

  port 11162 cannot be opened

Execution halted
```

**Comment by bgoodri** [comment]
I agree that forking would be preferable on a Linux server, although that is the least common way Stan is being used. It is easy enough to accomplish this yourself with something like
```
library(parallel)
library(rstan)
sm <- stan_model("foo.stan")
chains <- mclapply(1:chains, mc.cores = chains, mc.preschedule = FALSE, FUN = function(i) {
  sampling(sm, data = data_list, chains = 1, seed = 12345, chain_id = i)
})
post <- sflist2stanfit(chains)
```

**Comment by wookietreiber** [comment]
Wouldn't it just be easier for users (who are working regularly on Linux) to make the cluster an argument of the `stan` function, albeit with the default you are using at the moment?

**Comment by paul-buerkner** [comment]
I agree with @wookietreiber. For instance, I maintain a `cluster` argument in brms in addition to the much more elegant `cores` argument of rstan just to allow users to specify the cluster type. I would be very happy to only use `cores` going forward.

**Comment by bgoodri** [comment]
There is a feature/fork branch now that calls `mclapply` in the case where R is called from a shell on non-Windows computers. Calling `mclapply` from RStudio seems safe but would not print progress anywhere visible. So, you still have the old behavior with RStudio Server.

**Comment by bgoodri** [comment]
This was merged by 47d0c7d
